Add speed in -R json option (#4226)

wwhurin opened the pull request:
### Requirements

* Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* All new code requires tests to ensure against regressions.

### Description of the Change

I modified the ` -R json` option to show speed.

<!--

We must be able to understand the design of your change from this description. Keep in mind that the maintainers and/or community members reviewing this PR may not be familiar with the subsystem. Please be verbose.

-->

### Alternate Designs

N/A

<!-- Explain what other alternates were considered and why the proposed version was selected -->

### Why should this be in core?

#4226 

<!-- Explain why this functionality should be in mocha as opposed to its own package -->

### Benefits

You can see speed result in `json reporter `

<!-- What benefits will be realized by the code change? -->

### Possible Drawbacks

N/A

<!-- What are the possible side-effects or negative impacts of the code change? -->

### Applicable issues
Fixes #4226 

<!--
* Enter any applicable Issues here.

* Mocha follows semantic versioning: http://semver.org

* Is this a breaking change (major release)?
* Is it an enhancement (minor release)?
* Is it a bug fix, or does it not impact production code (patch release)?
-->

boneskull reviewed (changes requested):
Thanks!  Please update `test/reporters.json.spec.js` to reflect this change.  You would likely modify an existing assertion, e.g.,

```js
      expect(runner, 'to satisfy', {
        testResults: {
          failures: [
            {
              title: testTitle,
              file: testFile,
              speed: expect.it('to be greater than', 0),
              err: {
                message: error.message
              }
            }
          ]
        }
      });
```

outsideris reviewed (changes requested):
Could add this in `lib/reporters/json-stream.js` as well? 
And it needs to change `test/reporters/json-stream.spec.js` like boneskull said.

wwhurin commented:
> Thanks! Please update `test/reporters.json.spec.js` to reflect this change. You would likely modify an existing assertion, e.g.,
> 
> ```js
>       expect(runner, 'to satisfy', {
>         testResults: {
>           failures: [
>             {
>               title: testTitle,
>               file: testFile,
>               speed: expect.it('to be greater than', 0),
>               err: {
>                 message: error.message
>               }
>             }
>           ]
>         }
>       });
> ```

I have changed json.spec.js! However, speed is not a number, but things like 'fast' and'slow'. Also, speed does not pass unless it passes. But json.spec.js does not have a pass case, so I added a pass case to check the speed option. Would it be okay? 🤔

wwhurin commented:
> Could add this in `lib/reporters/json-stream.js` as well?
> And it needs to change `test/reporters/json-stream.spec.js` like boneskull said.

OK. I changed `json-stream.js` and `json-stream.spec.js`! You can see the speed in json-stream option.

outsideris reviewed (approved):
Thank  you.

boneskull commented on test/reporters/json.spec.js:
```diff
@@ -84,6 +85,32 @@ describe('JSON reporter', function() {
   });
 
+  it('should have 1 test pass', function(done) {
+    var test = new Test(testTitle, function(done) {
+      done();
+    });
```
```suggestion
    const test = new Test(testTitle, () => {});
```

boneskull commented on test/reporters/json.spec.js:
```diff
@@ -84,6 +85,32 @@ describe('JSON reporter', function() {
+            {
+              title: testTitle,
+              file: testFile,
+              speed: testSpeed
```
It's highly unlikely that `speed` would not be `fast` here, but since we haven't mocked the runner (as in `json-stream.spec.js`), it would be safer to do this:

```suggestion
              speed: /(slow|medium|fast)/
```

Also, you can remove the addition on [L14](https://github.com/mochajs/mocha/pull/4434/files#diff-1fa46f0ad13d99b90fbf14520f5b565eR14).

boneskull reviewed (approved):
Thanks for updating this.  I have a few suggestions.  If you don't apply them, I will do so after merge.

wwhurin commented on test/reporters/json.spec.js:
```diff
@@ -84,6 +85,32 @@ describe('JSON reporter', function() {
+    suite.addTest(test);
+
+    runner.run(function(failureCount) {
+      sinon.restore();
```
That's my fault. I didn't know exactly what role it is. 😂 Thank you!
